helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -2136,4 +2146,24 @@ public int hashCode() {
       return _listener.hashCode();
     }
   }
+
+  private void validateCurrentThread() {
+    if (_zookeeperEventThread != null && Thread.currentThread() == _zookeeperEventThread) {
+      throw new IllegalArgumentException("Must not be done in the zookeeper event thread.");
+    }
+  }
+
+  private void waitForKeeperStateUntilDeadline(KeeperState keeperState, Date deadline) {
```

Comment:
Why not put the lock/check current thread logic inside this method as well?

Reply:
Synced with @jiajunwang offline. Either works OK. We agree to add a locking layer for returning session, so there are less code changes and it is easier to maintain.